File: gst-python-samples/Gstreamer-Applications/gst_camera/usbcam_display.py
#!/usr/bin/env python3
import sys
import os
import gi
import logging

gi.require_version("GLib", "2.0")
gi.require_version("GObject", "2.0")
gi.require_version("Gst", "1.0")

from gi.repository import Gst, GLib

logging.basicConfig(level=logging.DEBUG,
                    format="[%(name)s] [%(levelname)8s] - %(message)s")
logger = logging.getLogger(__name__)


if __name__ == "__main__":

    Gst.init(sys.argv)
    loop = GLib.MainLoop()

    if len(sys.argv) != 2:
        print("Usage: %s </dev/videoX>", sys.argv[0])
        sys.exit(1)

    # creating empty pipeline
    pipeline = Gst.Pipeline.new("video-display")

    # creating elements
    source = Gst.ElementFactory.make("v4l2src",      "v4l2-source")
    videoconv = Gst.ElementFactory.make("videoconvert", "video-convert")
    videorate = Gst.ElementFactory.make("videorate",    "video-rate")
    framefilter = Gst.ElementFactory.make("capsfilter",   "frame-filter")
    sink = Gst.ElementFactory.make("waylandsink",  "display")

    # videoconvert: Convert video frames between a great variety of video formats.
    # we add videoconvert to convert the format from different usb camera input.

    if not pipeline or not source or not videoconv or not videorate or not framefilter or not sink:
        logger.error("Create element failed")
        sys.exit(1)

    source.set_property("device", sys.argv[1])

     # Modify the properties for framefilter
    framefilter.set_property("caps",
                             Gst.caps_from_string("video/x-raw,framerate=15/1,width=1280,height=720"))

     # Modify the properties for sink
    sink.set_property("x", 200)
    sink.set_property("y", 200)
    sink.set_property("width", 1280)
    sink.set_property("height", 720)

    bus = pipeline.get_bus()
    bus_id = bus.add_signal_watch()
    Gst.Object.unref(bus)

    #Build the pipeline
    pipeline.add(source)
    pipeline.add(videoconv)
    pipeline.add(videorate)
    pipeline.add(framefilter)
    pipeline.add(sink)

    if not source.link(videoconv):
        print("ERROR: Could not link source to framefilter")
        sys.exit(1)
    if not videoconv.link(videorate):
        print("ERROR: Could not link framefilter to sink")
        sys.exit(1)
    if not videorate.link(framefilter):
        print("ERROR: Could not link source to framefilter")
        sys.exit(1)
    if not framefilter.link(sink):
        print("ERROR: Could not link framefilter to sink")
        sys.exit(1)

try:
    # Start playing
    pipeline.set_state(Gst.State.PLAYING)
    print("Started Liveview")
    print('Press Ctrl+C to stop')
    loop.run()
except KeyboardInterrupt:
    print("Liveview stopped")
    pipeline.set_state(Gst.State.NULL)
    sys.exit(0)

File: RB5/linux_kernel_4_x/GStreamer-apps/cpp/gst_camera/usbcam_display.c
/*
* Copyright (c) 2020 Qualcomm Innovation Center, Inc.  All Rights Reserved.
*
* SPDX-License-Identifier: BSD-3-Clause-Clear
*/

/*
 * usbcam_display.c:
 *
 * A sample app based on gstreamer
 * The purpose is helping users to learn how to implement the
 * USB camera display on weston with gstreamer on the
 * Qualcomm platform through this sample app.
 *
 * This sample should be run in weston display.
 */

#include <gst/gst.h>
#include <glib.h>


static gboolean bus_callback(GstBus *bus, GstMessage *msg, gpointer data)
{
    GMainLoop *loop = (GMainLoop *)data;

    int type = GST_MESSAGE_TYPE(msg);

    if (type == GST_MESSAGE_ERROR) {
        gchar  *debug;
        GError *error;
        gst_message_parse_error(msg, &error, &debug);
        g_free(debug);
        g_printerr("Error: %s\n", error->message);
        g_error_free(error);
        g_main_loop_quit(loop);
    }

    return TRUE;
}


int main(int argc, char *argv[])
{
    GMainLoop *loop;

    GstElement *pipeline, *source, *videoconv, *videorate, *framefilter, *sink;
    GstBus *bus;
    guint bus_id;

    gst_init (&argc, &argv);

    loop = g_main_loop_new(NULL, FALSE);

    if (argc != 2) {
        g_printerr("Usage: %s </dev/videoX>\n", argv[0]);
        return -1;
    }

    pipeline    = gst_pipeline_new("video-display");
    source      = gst_element_factory_make("v4l2src",      "v4l2-source");
    videoconv   = gst_element_factory_make("videoconvert", "video-convert");
    videorate   = gst_element_factory_make("videorate",    "video-rate");
    framefilter = gst_element_factory_make("capsfilter",   "frame-filter");
    sink        = gst_element_factory_make("waylandsink",  "display");

    // videoconvert: Convert video frames between a great variety of video formats.
    // we add videoconvert to convert the format from different usb camera input.


    if (!pipeline || !source || !videoconv || !videorate || !framefilter || !sink) {
        g_printerr ("Create element failed.\n");
        return -1;
    }

    g_object_set(G_OBJECT(source), "device", argv[1], NULL);

    g_object_set(G_OBJECT(framefilter), "caps", 
                        gst_caps_from_string("video/x-raw,framerate=15/1,width=1280,height=720"), NULL);

    g_object_set (G_OBJECT(sink), "x", 0, "y", 100, "width", 640, "height", 360, NULL);

    bus = gst_pipeline_get_bus (GST_PIPELINE(pipeline));
    bus_id = gst_bus_add_watch(bus, bus_callback, loop);
    gst_object_unref(bus);

    gst_bin_add_many(GST_BIN(pipeline), source, videoconv, videorate, framefilter, sink, NULL);
    gst_element_link_many(source, videoconv, videorate, framefilter, sink, NULL);

    gst_element_set_state(pipeline, GST_STATE_PLAYING);
    g_print("Start\n");
    g_main_loop_run(loop);
    g_print("Stop\n");
    gst_element_set_state(pipeline, GST_STATE_NULL);

    gst_object_unref(GST_OBJECT(pipeline));
    g_source_remove(bus_id);
    g_main_loop_unref(loop);

    gst_deinit();

    return 0;
}
